docs: Pure Components (#1064)
deephaven/deephaven-plugins: plugins/ui/docs/describing/pure_components.md

# Pure Components

> A [pure function](https://en.wikipedia.org/wiki/Pure_function) returns the same value given the same arguments and has no side effects. By writing `deephaven.ui` components as pure functions, you can avoid bugs and unpredictable behavior.

Author: It is part of the definition that it has no side effect. If you write a function with side effects, then you have not written a pure function.

> A [pure function](https://en.wikipedia.org/wiki/Pure_function) returns the same value given the same arguments and has no side effects. By writing `deephaven.ui` components as pure functions, you can avoid bugs and unpredictable behavior.

Reviewer: I'm reading as an English person and not a programmer - you can dismiss my comment.

> ## Unintentional Side effects

Reviewer: ```suggestion
## Unintentional side effects
```

@ui.component
def tea_set():
    return ui.flex(cup(), cup(), cup(), direction="column")

my_tea_set1 = tea_set()
my_tea_set2 = tea_set()
```

![side effects](../_assets/pure_components1.png)

This component is reading and writing a `guest` variable declared outside of it. This means that calling this component multiple times will produce different results. If other components read `guest`, they will produce different results, too, depending on when they are rendered. That is not predictable.

You can fix this component by passing `guest` as a prop instead:

```python
from deephaven import ui

@ui.component
def cup(guest):
    return ui.text(f"Tea cup for guest {guest}")

@ui.component
def tea_set():
    return ui.flex(cup(guest=1), cup(guest=2), cup(guest=3), direction="column")

my_tea_set1 = tea_set()
my_tea_set2 = tea_set()
```

![side effects 2](../_assets/pure_components2.png)

Now the component is pure. It returns only depends on the `guest` prop.

In general, you should not expect components to be rendered in any particular order. Each component should only “think for itself”, and not attempt to coordinate with or depend upon others during rendering.

> ## Local Mutations

Reviewer: ```suggestion
## Local mutations
```

Pure functions do not mutate variables outside of the function's scope or objects that were created before the function call. However, it is fine to change variables and objects created inside the function. In this example, the component creates list and adds a dozen cups to it:

```python
from deephaven import ui

@ui.component
def cup(guest):
    return ui.text(f"Tea cup for guest {guest}")

@ui.component
def tea_set():
    cups = []
    for i in range(1, 13):
        cups.append(cup(guest=i))
    return ui.flex(cups, direction="column")

my_tea_set1 = tea_set()
my_tea_set2 = tea_set()
```

![local mutations](../_assets/pure_components3.png)

> If the `cups` variable was outside the TeaGathering function, this would be a problem. You would be changing a preexisting object by appending items to that list.

Reviewer 2: ```suggestion
If the `cups` variable was outside the `tea_set` function, this would be a problem. You would be changing a preexisting object by appending items to that list.
```

> However, because you created them during the same render, no code outside of `tea_set` will be impacted by this. This a local mutation.

Reviewer: ```suggestion
However, because you created them during the same render, no code outside of `tea_set` will be impacted by this. This is a local mutation.
```

> ## Intentional Side Effects

Reviewer: ```suggestion
## Intentional side effects
```

> While the rendering process must remain pure, at some point, something needs to change. You may need to print a message, update the screen, start an animation, or change data. These changes are called side effects. These things need to happen on the side rather than during rendering.

Reviewer: ```suggestion
While the rendering process must remain pure, at some point, something needs to change. You may need to print a message, update the screen, start an animation, or change data. These changes are called side effects. They must happen on the side rather than during rendering.
```

In `deephaven.ui`, side effects usually belong in event handlers. Event handlers are functions that run when you perform some action like clicking a button. Even though the event handlers are defined inside your component, they do not run during rendering, so even handlers do not need to be pure.

```python
from deephaven import ui

@ui.component
def event_handler_example():
    # An event handler for a button
    def button_handler():

>         print("button pressed")

Reviewer 2: In the opening paragraph of this section, you say "update the screen, start an animation, or change data", then in this example you're just printing a log message. I think it would be more powerful to match the two; either include "print a message" in the opening paragraph, or change this example to update some data.

Author: Updated

    return ui.button("button", on_press=button_handler)

my_event_handler_example = event_handler_example()
```

> If an event handler is not the correct place for a certain side effect, you can place it in a [`use_effect`](../hooks/use_effect.md) hook. This tells `deephaven.ui` to execute it later, after rendering, when side affects are allowed.

Reviewer: ```suggestion
If an event handler is not the correct place for a certain side effect, you can place it in a [`use_effect`](../hooks/use_effect.md) hook. This tells `deephaven.ui` to execute it later, after rendering, when side effects are allowed.
```
